plugins/outputs/sematext/processors/metainfo.go

```diff
@@ -0,0 +1,248 @@
+	return &Metainfo{
+		log:          log,
+		token:        token,
+		sentMetrics:  sentMetricsMap,
+		sendChannel:  make(chan bool, 1),
+		metainfoURL:  receiverURL + "/write?db=metainfo",
+		senderConfig: senderConfig,
+		sender:       sender.NewSender(senderConfig),
+		serializer:   NewMetainfoSerializer(log),
+	}
+}
+
+// Process contains core logic of Metainfo processor
+func (m *Metainfo) Process(metrics []telegraf.Metric) ([]telegraf.Metric, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	newMetrics := make(map[string]*MetricMetainfo)
+
+	for _, metric := range metrics {
+		for _, field := range metric.FieldList() {
+			mInfo, mKey := processMetric(m.token, &metric, field, &m.sentMetrics)
+			if mInfo != nil {
+				newMetrics[mKey] = mInfo
+			}
+		}
+	}
+
+	if len(newMetrics) > 0 {
+		go m.sendMetainfo(newMetrics)
+
+		// add newMetrics to sentMetrics so they are not sent again from other goroutines
+		for k, v := range newMetrics {
+			m.sentMetrics[k] = v
+		}
+
+	}
+
+	return metrics, nil
+}
+
+func (m *Metainfo) sendMetainfo(newMetrics map[string]*MetricMetainfo) {
+	reqCounter := 0
+
+	mInfoSlice := make([]*MetricMetainfo, len(newMetrics))
+	for _, mInfo := range newMetrics {
+		mInfoSlice = append(mInfoSlice, mInfo)
+	}
+	body := m.serializer.Write(mInfoSlice)
+
+	for {
+		select {
+		case <-m.sendChannel:
+			return
+		default:
+			m.log.Infof("sending metainfo to Sematext endpoint %s", m.metainfoURL)
+			res, err := m.sender.Request("POST", m.metainfoURL, "text/plain; charset=utf-8", body)
+			reqCounter++
+			if err != nil {
+				// possibly non-recoverable, return without retrying
+				m.log.Errorf("can't send metainfo to Sematext endpoint %s, error: %v",
+					m.metainfoURL, err)
+				return
+			}
+			responseContent := response(res)
+			res.Body.Close()
+
+			success := res.StatusCode >= 200 && res.StatusCode < 300
+
+			if success {
+				m.log.Infof("successfully sent metainfo to Sematext endpoint %s, batch size : %d",
+					m.metainfoURL, len(newMetrics))
+				return
+			}
+
+			m.log.Errorf("can't send the metainfo to Sematext endpoint %s, response code: %d, response: %s",
+				m.metainfoURL, res.StatusCode, responseContent)
+
+			badRequest := res.StatusCode >= 400 && res.StatusCode < 500
+			if badRequest {
+				m.log.Infof("no retry for bad requests, response code was %d", res.StatusCode)
+				return
+			}
+
+			if reqCounter >= maxMetainfoRetries {
+				m.log.Warnf("max retries (%d) exceeded, cancelling the request permanently",
+					maxMetainfoRetries)
+				return
+			}
+
+			nextSleepIntervalSec := int32(reqCounter * metainfoRetryIntervalSeconds)
+			m.log.Infof("metainfo sending retry in %d seconds", nextSleepIntervalSec)
+			time.Sleep(time.Second * time.Duration(rand.Int31n(nextSleepIntervalSec)))
+		}
+	}
+}
+
+func processMetric(token string, metric *telegraf.Metric, field *telegraf.Field,
+	sentMetrics *map[string]*MetricMetainfo) (*MetricMetainfo, string) {
+	host, set := (*metric).GetTag(telegrafHostTag)
```

Comment:
(*metric).GetTag(telegrafHostTag) -> metric.GetTag(telegrafHosTag) ?

Reply:
Was passing an interface by reference, fixed it now.